README.md
# Flutter Date Picker Timeline
![Flutter Date Picker Timeline Banner](https://raw.githubusercontent.com/sobimor/flutter_date_picker_timeline/master/repo_files/images/banner.png)
[![pub package](https://img.shields.io/pub/v/flutter_date_picker_timeline.svg)](https://pub.dartlang.org/packages/flutter_date_picker_timeline)

💥 Gregorian and Jalali customizable date picker as a horizontal timeline  💥

<p align="center">
<img src="https://raw.githubusercontent.com/sobimor/flutter_date_picker_timeline/master/repo_files/images/screenshot_ios.png" width="300" alt="Screenshot iOS">
&nbsp;&nbsp;&nbsp;&nbsp;
<img src="https://raw.githubusercontent.com/sobimor/flutter_date_picker_timeline/master/repo_files/images/screenshot_android.jpg" width="300" alt="Screenshot android">
</p><br>

# Let's get started

### 1 - Depend on it

##### Add it to your package's pubspec.yaml file

```yml
dependencies:
  flutter_date_picker_timeline: ^0.3.3
```


### 2 - Install it

##### Install packages from the command line
```sh
flutter pub get
```


### 3 - Import it

##### Import it to your project
```dart
import 'package:flutter_date_picker_timeline/flutter_date_picker_timeline.dart';
```


## How to use?
##### Use the `FlutterDatePickerTimeline` Widget

```dart
FlutterDatePickerTimeline(
    startDate: DateTime(2020, 07, 01),
    endDate: DateTime(2020, 12, 30),
    initialSelectedDate: DateTime(2020, 07, 24),
    onSelectedDateChange: (DateTime dateTime) {
        print(dateTime);
      },
    )
```

##### Constructor:

```dart
FlutterDatePickerTimeline({
    Key key,
    // Determines the [FlutterDatePickerTimeline] mode.
    this.calendarMode = CalendarMode.gregorian,
    // The first date of [FlutterDatePickerTimeline].
    @required this.startDate,
    // The last date of [FlutterDatePickerTimeline].
    @required this.endDate,
    // Initially selected date.
    this.initialSelectedDate,
    // Initially focused date(If nothing is provided, a [initialSelectedDate] will be used).
    this.initialFocusedDate,
    // Used for setting the textDirection of [FlutterDatePickerTimeline].
    this.textDirection,
    // Used for setting the width of selected items.
    this.selectedItemWidth = 170,
    // Used for setting the width of unselected items.
    this.unselectedItemWidth = 38,
    // Used for setting the height of selected and unselected items.
    this.itemHeight = 38,
    // Used for setting the radius of selected and unselected items background.
    this.itemRadius = 10,
    // Used for setting the padding of [ListView].
    this.listViewPadding = const EdgeInsets.only(right: 5.5, left: 5.5),
    // Used for setting the margin of selected items.
    this.selectedItemMargin = const EdgeInsets.only(right: 5.5, left: 5.5),
    // Used for setting the margin of unselected items.
    this.unselectedItemMargin = const EdgeInsets.only(right: 5.5, left: 5.5),
    // Used for setting the color of selected items background.
    this.selectedItemBackgroundColor = const Color(0xFF2B2C30),
    // Used for setting the color of unselected items background.
    this.unselectedItemBackgroundColor = Colors.white,
    // Used for setting the style of selected items [Text].
    this.selectedItemTextStyle,
    // Used for setting the style of unselected items [Text].
    this.unselectedItemTextStyle,
    // Called whenever any date gets selected.
    @required this.onSelectedDateChange,
  }) : super(key: key);
```

## Props
| props                   | types           | defaultValues                                                                                                     |
| :---------------------- | :-------------: | :---------------------------------------------------------------------------------------------------------------: |
| calendarMode        | `CalendarMode`        | CalendarMode.gregorian |
| startDate       | `DateTime`     | |
| endDate           | `DateTime`     | |
| initialSelectedDate       | `DateTime?`     | |
| initialFocusedDate | `DateTime?`         | If nothing is provided, a [initialSelectedDate] will be used |
| textDirection       | `TextDirection?`     | If nothing is provided, a [calendarMode] will be used. (CalendarMode.gregorian -> TextDirection.ltr , CalendarMode.jalali -> TextDirection.rtl) |
| selectedItemWidth       | `double`     | 170 |
| unselectedItemWidth | `double`         | 38 |
| itemHeight       | `double`     | 38 |
| itemRadius | `double`         | 10 |
| listViewPadding       | `EdgeInsets`     | const EdgeInsets.only(right: 5.5, left: 5.5) |
| selectedItemMargin | `EdgeInsets`         | const EdgeInsets.only(right: 5.5, left: 5.5) |
| unselectedItemMargin       | `EdgeInsets`     | const EdgeInsets.only(right: 5.5, left: 5.5) |
| selectedItemBackgroundColor | `Color`         | const Color(0xFF2B2C30) |
| unselectedItemBackgroundColor       | `Color`     | Colors.white |
| selectedItemTextStyle | `TextStyle?`         | TextStyle(fontFamily: widget.calendarMode == CalendarMode.gregorian ? 'nunito' : 'dana', package: 'flutter_date_picker_timeline', color: widget.unselectedItemBackgroundColor) |
| unselectedItemTextStyle       | `TextStyle?`     | TextStyle(fontFamily: widget.calendarMode == CalendarMode.gregorian ? 'nunito' : 'dana', package: 'flutter_date_picker_timeline', color: widget.selectedItemBackgroundColor) |
| onSelectedDateChange | `DateChangeListener`         | |

CHANGELOG.md
## 0.3.3
* fix: update shamsi_date, scroll_to_index plugins to the last version.

## 0.3.2
* fix: update shamsi_date plugin to the last version.

## 0.3.1
* fix: update shamsi_date plugin to the last version.

## 0.3.0
* **Breaking**: migrated to null safety.

## 0.2.0
* feat: now startDate and endDate can be same for showing just one date, #3.
* fix: now initialSelectedDate can change after initial set, #3.

## 0.1.6
* feat: add textDirection property in [FlutterDatePickerTimeline](https://github.com/sobimor/flutter_date_picker_timeline/blob/master/lib/src/flutter_date_picker_timeline_widget.dart), #2.

## 0.1.5
* fix: downgrade intl plugin to the last stable version.

## 0.1.4
* fix: update intl plugin to the last version.

## 0.1.3
* fix: selecting multiple selected dates.

## 0.1.2
* formatted according to dartfmt.

## 0.1.1
* formatted according to dartfmt.

## 0.1.0 - Released on (2020 Dec 2).

example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:flutter_date_picker_timeline_example/main.dart';

void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MyApp());

    // Verify that our counter starts at 0.
    expect(find.text('0'), findsOneWidget);
    expect(find.text('1'), findsNothing);

    // Tap the '+' icon and trigger a frame.
    await tester.tap(find.byIcon(Icons.add));
    await tester.pump();

    // Verify that our counter has incremented.
    expect(find.text('0'), findsNothing);
    expect(find.text('1'), findsOneWidget);
  });
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:flutter_date_picker_timeline/flutter_date_picker_timeline.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter date picker timeline example',
      theme: ThemeData(
        primarySwatch: Colors.blueGrey,
      ),
      home: Scaffold(
          backgroundColor: Colors.white,
          body: Center(
            child: Container(
                padding: const EdgeInsets.only(top: 11, bottom: 11),
                decoration: BoxDecoration(color: const Color(0xFFF5F5F5)),
                child: FlutterDatePickerTimeline(
                  startDate: DateTime(2020, 07, 01),
                  endDate: DateTime(2020, 12, 30),
                  initialSelectedDate: DateTime(2020, 07, 24),
                  onSelectedDateChange: (DateTime dateTime) {
                    print(dateTime);
                  },
                )),
          )),
    );
  }
}


lib/src/utils/date_helper_extension.dart
import 'package:intl/intl.dart';
import 'package:shamsi_date/shamsi_date.dart';

extension DateHelperExtension on DateTime {
  int dateDifference(DateTime secondDate) {
    return DateTime(this.year, this.month, this.day)
        .difference(DateTime(secondDate.year, secondDate.month, secondDate.day))
        .inDays;
  }

  bool isSameDate(DateTime other) {
    return this.year == other.year &&
        this.month == other.month &&
        this.day == other.day;
  }

  bool isInRange(DateTime startDate, DateTime endDate) {
    return (this.isAfter(startDate) || this.isSameDate(startDate)) &&
        (this.isBefore(endDate) || this.isSameDate(endDate));
  }

  String getGregorianWeekDayAndDate() {
    final f = DateFormat('EEEE, MMM d');

    return f.format(this);
  }

  String getJalaliDay() {
    final f = Jalali.fromDateTime(this).formatter;

    return '${f.d}';
  }

  String getJalaliWeekDayAndDate() {
    final f = Jalali.fromDateTime(this).formatter;

    return '${f.wN}، ${f.d} ${f.mN}';
  }
}


lib/src/widgets/item/date_picker_item_selected.dart
import 'package:flutter/material.dart';
import 'package:flutter_date_picker_timeline/src/utils/date_helper_extension.dart';
import 'package:flutter_date_picker_timeline/src/models/enums/calendar_mode.dart';

class DatePickerItemSelected extends StatelessWidget {
  const DatePickerItemSelected(
      {required this.date,
      required this.calendarMode,
      required this.width,
      required this.itemRadius,
      required this.itemMargin,
      required this.itemBackgroundColor,
      required this.textStyle});

  final DateTime date;
  final CalendarMode calendarMode;
  final double width;
  final double itemRadius;
  final EdgeInsets itemMargin;
  final Color itemBackgroundColor;
  final TextStyle textStyle;

  @override
  Widget build(BuildContext context) {
    return Container(
        width: width,
        margin: itemMargin,
        decoration: new BoxDecoration(
            color: itemBackgroundColor,
            borderRadius: BorderRadius.all(Radius.circular(itemRadius))),
        child: Center(
            child: Text(
          calendarMode == CalendarMode.gregorian
              ? date.getGregorianWeekDayAndDate()
              : date.getJalaliWeekDayAndDate(),
          style: textStyle,
        )));
  }
}


lib/src/widgets/item/date_picker_item_unselected.dart
import 'package:flutter/material.dart';
import 'package:flutter_date_picker_timeline/src/models/enums/calendar_mode.dart';
import 'package:flutter_date_picker_timeline/src/utils/date_helper_extension.dart';

class DatePickerItemUnselected extends StatelessWidget {
  const DatePickerItemUnselected(
      {required this.date,
      required this.calendarMode,
      required this.width,
      required this.itemRadius,
      required this.itemMargin,
      required this.itemBackgroundColor,
      required this.textStyle,
      required this.onPressed});

  final DateTime date;
  final CalendarMode calendarMode;
  final double width;
  final double itemRadius;
  final EdgeInsets itemMargin;
  final Color itemBackgroundColor;
  final TextStyle textStyle;
  final VoidCallback onPressed;

  @override
  Widget build(BuildContext context) {
    return GestureDetector(
      onTap: onPressed,
      child: Container(
          width: width,
          margin: itemMargin,
          decoration: new BoxDecoration(
              color: itemBackgroundColor,
              borderRadius: BorderRadius.all(Radius.circular(itemRadius))),
          child: Center(
              child: Text(
            calendarMode == CalendarMode.gregorian
                ? date.day.toString()
                : date.getJalaliDay(),
            style: textStyle,
          ))),
    );
  }
}


lib/src/flutter_date_picker_timeline_widget.dart
import 'package:flutter/material.dart';
import 'package:flutter_date_picker_timeline/src/models/enums/calendar_mode.dart';
import 'package:flutter_date_picker_timeline/src/utils/ui/custom_scroll_behavior.dart';
import 'package:flutter_date_picker_timeline/src/widgets/item/date_picker_item_unselected.dart';
import 'package:flutter_date_picker_timeline/src/widgets/item/date_picker_item_selected.dart';
import 'package:flutter_date_picker_timeline/src/utils/date_helper_extension.dart';
import 'package:scroll_to_index/scroll_to_index.dart';

/// Callback exposing currently selected date.
typedef DateChangeListener = Function(DateTime? date);

/// Gregorian and Jalali customizable date picker as a horizontal timeline
class FlutterDatePickerTimeline extends StatefulWidget {
  FlutterDatePickerTimeline({
    Key? key,
    this.calendarMode = CalendarMode.gregorian,
    required this.startDate,
    required this.endDate,
    this.initialSelectedDate,
    this.initialFocusedDate,
    this.textDirection,
    this.selectedItemWidth = 170,
    this.unselectedItemWidth = 38,
    this.itemHeight = 38,
    this.itemRadius = 10,
    this.listViewPadding = const EdgeInsets.only(right: 5.5, left: 5.5),
    this.selectedItemMargin = const EdgeInsets.only(right: 5.5, left: 5.5),
    this.unselectedItemMargin = const EdgeInsets.only(right: 5.5, left: 5.5),
    this.selectedItemBackgroundColor = const Color(0xFF2B2C30),
    this.unselectedItemBackgroundColor = Colors.white,
    this.selectedItemTextStyle,
    this.unselectedItemTextStyle,
    required this.onSelectedDateChange,
  })  : assert((endDate.isSameDate(startDate) || endDate.isAfter(startDate)),
            "The end date must be same or before the start date!"),
        assert(
            initialSelectedDate == null ||
                initialSelectedDate.isInRange(startDate, endDate),
            "The initialSelectedDate must be in the start and end date range!"),
        assert(
            initialFocusedDate == null ||
                initialFocusedDate.isInRange(startDate, endDate),
            "The initialFocusedDate must be in the start and end date range!"),
        super(key: key);

  /// Determines the [FlutterDatePickerTimeline] mode.
  ///
  /// Defaults to [CalendarMode.gregorian]
  final CalendarMode calendarMode;

  /// The first date of [FlutterDatePickerTimeline].
  final DateTime startDate;

  /// The last date of [FlutterDatePickerTimeline].
  final DateTime endDate;

  /// Initially selected date. Usually it will be `DateTime.now()`.
  final DateTime? initialSelectedDate;

  /// Initially focused date. Usually it will be `DateTime.now()`.
  ///
  /// If nothing is provided, a [initialSelectedDate] will be used.
  final DateTime? initialFocusedDate;

  /// Used for setting the textDirection of [FlutterDatePickerTimeline].
  ///
  /// If nothing is provided, a [calendarMode] will be used.
  /// For [CalendarMode.gregorian] use [TextDirection.ltr]
  /// For [CalendarMode.jalali] use [TextDirection.rtl]
  final TextDirection? textDirection;

  /// Used for setting the width of selected items.
  ///
  /// Defaults to 170
  final double selectedItemWidth;

  /// Used for setting the width of unselected items.
  ///
  /// Defaults to 38
  final double unselectedItemWidth;

  /// Used for setting the height of selected and unselected items.
  ///
  /// Defaults to 38
  final double itemHeight;

  /// Used for setting the radius of selected and unselected items background.
  ///
  /// Defaults to 10
  final double itemRadius;

  /// Used for setting the padding of [ListView].
  ///
  /// Defaults to [const EdgeInsets.only(right: 5.5, left: 5.5)]
  final EdgeInsets listViewPadding;

  /// Used for setting the margin of selected items.
  ///
  /// Defaults to [const EdgeInsets.only(right: 5.5, left: 5.5)]
  final EdgeInsets selectedItemMargin;

  /// Used for setting the margin of unselected items.
  ///
  /// Defaults to [const EdgeInsets.only(right: 5.5, left: 5.5)]
  final EdgeInsets unselectedItemMargin;

  /// Used for setting the color of selected items background.
  ///
  /// Defaults to [const Color(0xFF2B2C30)]
  final Color selectedItemBackgroundColor;

  /// Used for setting the color of unselected items background.
  ///
  /// Defaults to [Colors.white]
  final Color unselectedItemBackgroundColor;

  /// Used for setting the style of selected items [Text].
  ///
  /// Defaults to [TextStyle(
  ///       fontFamily: widget.calendarMode == CalendarMode.gregorian ? 'nunito' : 'dana',
  ///       package: 'flutter_date_picker_timeline',
  ///       color: widget.unselectedItemBackgroundColor)]
  final TextStyle? selectedItemTextStyle;

  /// Used for setting the style of unselected items [Text].
  ///
  /// Defaults to [TextStyle(
  ///       fontFamily: widget.calendarMode == CalendarMode.gregorian ? 'nunito' : 'dana',
  ///       package: 'flutter_date_picker_timeline',
  ///       color: widget.selectedItemBackgroundColor)]
  final TextStyle? unselectedItemTextStyle;

  /// Called whenever any date gets selected.
  final DateChangeListener onSelectedDateChange;

  @override
  _FlutterDatePickerTimelineState createState() =>
      _FlutterDatePickerTimelineState();
}

class _FlutterDatePickerTimelineState extends State<FlutterDatePickerTimeline> {
  late AutoScrollController _scrollController;
  late ValueNotifier<DateTime?> _selectedDateValueNotifier;

  get _defaultSelectedItemTextStyle => TextStyle(
      fontFamily:
          widget.calendarMode == CalendarMode.gregorian ? 'nunito' : 'dana',
      package: 'flutter_date_picker_timeline',
      color: widget.unselectedItemBackgroundColor);

  get _defaultUnselectedItemTextStyle => TextStyle(
      fontFamily:
          widget.calendarMode == CalendarMode.gregorian ? 'nunito' : 'dana',
      package: 'flutter_date_picker_timeline',
      color: widget.selectedItemBackgroundColor);

  @override
  void initState() {
    _scrollController = AutoScrollController(
      axis: Axis.horizontal,
    );
    _selectedDateValueNotifier =
        ValueNotifier<DateTime?>(widget.initialSelectedDate);
    if (widget.initialFocusedDate != null ||
        widget.initialSelectedDate != null) {
      _scrollToInitialFocusedDate();
    }
    super.initState();
  }

  @override
  void didUpdateWidget(covariant FlutterDatePickerTimeline oldWidget) {
    if (oldWidget.initialSelectedDate != widget.initialSelectedDate) {
      _setSelectedDate(widget.initialSelectedDate);
    }
    super.didUpdateWidget(oldWidget);
  }

  @override
  void dispose() {
    _selectedDateValueNotifier.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Directionality(
      textDirection: widget.textDirection ??
          (widget.calendarMode == CalendarMode.jalali
              ? TextDirection.rtl
              : TextDirection.ltr),
      child: Container(
        height: widget.itemHeight,
        child: ScrollConfiguration(
            behavior: CustomScrollBehavior(),
            child: ValueListenableBuilder<DateTime?>(
                valueListenable: _selectedDateValueNotifier,
                builder: (BuildContext context, DateTime? selectedDate,
                    Widget? child) {
                  return ListView.builder(
                      padding: widget.listViewPadding,
                      scrollDirection: Axis.horizontal,
                      controller: _scrollController,
                      itemCount:
                          widget.endDate.dateDifference(widget.startDate) + 1,
                      itemBuilder: (BuildContext context, int index) {
                        DateTime itemDate =
                            widget.startDate.add(Duration(days: index));
                        return AutoScrollTag(
                            key: ValueKey(index),
                            controller: _scrollController,
                            index: index,
                            child: (_isDatePickerItemSelected(
                                    itemDate, selectedDate))
                                ? DatePickerItemSelected(
                                    date: itemDate,
                                    calendarMode: widget.calendarMode,
                                    width: widget.selectedItemWidth,
                                    itemRadius: widget.itemRadius,
                                    itemMargin: widget.selectedItemMargin,
                                    itemBackgroundColor:
                                        widget.selectedItemBackgroundColor,
                                    textStyle: widget.selectedItemTextStyle ??
                                        _defaultSelectedItemTextStyle,
                                  )
                                : DatePickerItemUnselected(
                                    date: itemDate,
                                    calendarMode: widget.calendarMode,
                                    width: widget.unselectedItemWidth,
                                    itemRadius: widget.itemRadius,
                                    itemMargin: widget.unselectedItemMargin,
                                    itemBackgroundColor:
                                        widget.unselectedItemBackgroundColor,
                                    textStyle: widget.unselectedItemTextStyle ??
                                        _defaultUnselectedItemTextStyle,
                                    onPressed: () {
                                      _setSelectedDate(itemDate);
                                      _scrollToIndex(index);
                                    },
                                  ));
                      });
                })),
      ),
    );
  }

  bool _isDatePickerItemSelected(DateTime itemDate, DateTime? selectedDate) {
    if (selectedDate == null) return false;
    return itemDate.dateDifference(selectedDate) == 0;
  }

  _setSelectedDate(DateTime? date) {
    _selectedDateValueNotifier.value = date;
    widget.onSelectedDateChange(date);
  }

  _scrollToIndex(index) async {
    await _scrollController.scrollToIndex(index,
        preferPosition: AutoScrollPosition.middle);
  }

  _scrollToInitialFocusedDate() async {
    final DateTime initialFocusedDate =
        widget.initialFocusedDate ?? widget.initialSelectedDate!;
    _scrollToIndex(initialFocusedDate.dateDifference(widget.startDate));
  }
}
